app/services/timebox.rb

```diff
@@ -0,0 +1,21 @@
+# An object that runs a provided block within the time limit or else returns
+# the result of the fallback block. run may be called multiple times; the
+# original time limit being applied to all runs collectively.
+class Timebox
+  def initialize(time_limit)
+    @deadline = Time.now.to_i + time_limit
+  end
+
+  def run(fallback_block)
+    time_remaining = @deadline - Time.now.to_i
```

Comment:
Using `to_i` on the time will lead to very inconsistent timeouts being applied in practice, depending on the fractional component of the time when the deadline and current time are recorded.